internal/xds/translator/translator.go

```diff
@@ -627,6 +631,31 @@ func (t *Translator) processTCPListenerXdsTranslation(
 				errs = errors.Join(errs, err)
 			}
 		}
+
+		// If there are no routes, add a route without a destination to the listener to create a filter chain
+		// This is needed because Envoy requires a filter chain to be present in the listener, otherwise it will reject the listener and report a warning
+		if len(tcpListener.Routes) == 0 {
+			nullRouteCluster := &clusterv3.Cluster{
```

Comment:
hey curious, what happens if we set `xdsListener.FilterChains = []` for this case, does xDS still complain ?

Reply:
Envoy will complain `no filter chains specified`. It's a validation inside the Envoy listener impl instead of a proto type built-in validation.

https://github.com/envoyproxy/envoy/blob/79958991ffe0cf8d59a4a351646c76d672dada83/source/common/listener_manager/listener_impl.cc#L763-L770

```
type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) gateway-conformance-infra/same-namespace/tls: error adding listener '0.0.0.0:10443': no filter chains specified
```

Reply:
Envoy will complain `no filter chains specified`. It's a validation inside the Envoy listener impl instead of a proto type built-in validation.

https://github.com/envoyproxy/envoy/blob/79958991ffe0cf8d59a4a351646c76d672dada83/source/common/listener_manager/listener_impl.cc#L763-L770

```
type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) gateway-conformance-infra/same-namespace/tls: error adding listener '0.0.0.0:10443': no filter chains specified
```